test/stats.node.js

```diff
@@ -0,0 +1,205 @@
+      expect(err).to.not.exist()
+      const expectedTransports = [
+        '/mplex/6.7.0',
+        '/secio/1.0.0'
+      ]
+
+      switches.forEach(
+        (swtch) => expect(swtch.stats.transports().sort()).to.deep.equal(expectedTransports))
+      teardown(switches, done)
+    })
+  })
+
+  it('both nodes know the protocols', (done) => {
+    setup((err, switches) => {
+      expect(err).to.not.exist()
+      const expectedProtocols = [
+        '/echo/1.0.0',
+        '/mplex/6.7.0'
+      ]
+
+      switches.forEach((swtch) => {
+        expect(swtch.stats.protocols().sort()).to.deep.equal(expectedProtocols)
+      })
+
+      teardown(switches, done)
+    })
+  })
+
+  it('both nodes know about each other', (done) => {
+    setup((err, switches) => {
+      expect(err).to.not.exist()
+      switches.forEach(
+        (swtch, index) => {
+          const otherSwitch = selectOther(switches, index)
+          expect(swtch.stats.peers().sort()).to.deep.equal([otherSwitch._peerInfo.id.toB58String()])
+        })
+      teardown(switches, done)
+    })
+  })
+
+  it('both have transport-specific stats', (done) => {
+    setup((err, switches) => {
+      expect(err).to.not.exist()
+      switches.forEach((swtch) => {
+        let snapshot = swtch.stats.forTransport('/secio/1.0.0').snapshot
+        expect(snapshot.dataReceived.toFixed()).to.equal('92')
+        expect(snapshot.dataSent.toFixed()).to.equal('90')
+      })
+      teardown(switches, done)
+    })
+  })
+
+  it('both have protocol-specific stats', (done) => {
+    setup((err, switches) => {
+      expect(err).to.not.exist()
+      switches.forEach((swtch) => {
+        let snapshot = swtch.stats.forProtocol('/echo/1.0.0').snapshot
+        expect(snapshot.dataReceived.toFixed()).to.equal('4')
+        expect(snapshot.dataSent.toFixed()).to.equal('8')
+      })
+      teardown(switches, done)
+    })
+  })
+
+  it('both have peer-specific stats', (done) => {
+    setup((err, switches) => {
+      expect(err).to.not.exist()
+      switches.forEach((swtch, index) => {
+        const other = selectOther(switches, index)
+        let snapshot = swtch.stats.forPeer(other._peerInfo.id.toB58String()).snapshot
+        expect(snapshot.dataReceived.toFixed()).to.equal('51')
+        expect(snapshot.dataSent.toFixed()).to.equal('49')
+      })
+      teardown(switches, done)
+    })
+  })
+
+  it('both have moving average stats for peer', (done) => {
+    setup((err, switches) => {
+      expect(err).to.not.exist()
+      switches.forEach((swtch, index) => {
+        const other = selectOther(switches, index)
+        let ma = swtch.stats.forPeer(other._peerInfo.id.toB58String()).movingAverages
+        const intervals = [60000, 300000, 900000]
+        intervals.forEach((interval) => {
+          const average = ma.dataReceived[interval].movingAverage()
+          expect(average).to.be.above(0).below(1)
+        })
+      })
+      teardown(switches, done)
+    })
+  })
+})
```

Comment:
We need tests for:

- [ ] Clearing out previous stats from a peer (Rep mechanisms will need a reset)
- [ ] Check what happens when a peer disconnects and connects again
- [ ] Check when a peer disconnects and connects again from a different transport

Reply:
@diasdavid Currently, peer disconnections imply losing the stats for that peer.
If you think that it's important to keep them around, we should impose a time limit so that I eventually becomes garbage collected...
@diasdavid thoughts?

Reply:
Perhaps tracking the sessions and having them in a LRU cache?

Reply:
@diasdavid good call, but I'm not going to put all of the sessions there, only once a peer disconnects..